src/sagemaker_xgboost_container/algorithm_mode/handler_service.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+import numpy as np
+import os
+import pickle as pkl
+from scipy.sparse import csr_matrix
+import xgboost as xgb
+
+from sagemaker_inference import content_types, default_inference_handler, encoder
+from sagemaker_inference.default_handler_service import DefaultHandlerService
+
+from sagemaker_xgboost_container.algorithm_mode.mms_transformer import Transformer
+from sagemaker_xgboost_container import encoder as xgb_encoder
+from sagemaker_xgboost_container.algorithm_mode.inference_errors import NoContentInferenceError, \
+    UnsupportedMediaTypeInferenceError, ModelLoadInferenceError, BadRequestInferenceError
+
+
+SAGEMAKER_BATCH = os.getenv("SAGEMAKER_BATCH")
+
+
+# FIXME: https://github.com/aws/sagemaker-xgboost-container/issues/12
+# This was failing a previous container test; need to make decision as to change test or behavior.
+def _get_sparse_matrix_from_libsvm(payload):
+    pylist = map(lambda x: x.split(' '), payload.split('\n'))
+    colon = ':'
+    row = []
+    col = []
+    data = []
+    for row_idx, line in enumerate(pylist):
+        for item in line:
+            if colon in item:
+                col_idx = item.split(colon)[0]
+                val = item.split(colon)[1]
+                row.append(row_idx)
+                col.append(col_idx)
+                data.append(val)
+
+    row = np.array(row)
+    col = np.array(col).astype(np.int)
+    data = np.array(data).astype(np.float)
+    if not (len(row) == len(col) and len(col) == len(data)):
+        raise RuntimeError("Dimension checking failed when transforming sparse matrix.")
+
+    return csr_matrix((data, (row, col)))
+
+
+def predict(booster, model_format, dtest, content_type):
+    if model_format == 'pkl_format':
+        x = len(booster.feature_names)
+        y = len(dtest.feature_names)
+
+        if content_type == 'text/x-libsvm' or content_type == 'text/libsvm':
+            if y > x + 1:
```

Comment:
What's the context of `+1` here and in CSV? Is that because the labels could be included in the test data?

Reply:
To be honest, I'm not sure. This was taken from the legacy xgboost model and I never questioned it because it was passing container tests. Might be worth doing a deep dive.

Reply:
I also thought it might have to do with the label being included in the actual data of the DMatrix, but I'm not sure if that really makes sense. Locally I was able to confirm that there is an error if the feature names do not match. So the `y == x + 1` case doesn't seem like it should work.

Reply:
I tried running container tests after removing these checks, and it actually breaks the inference tests. Not sure why this works, but looks like it is necessary.